### ISSUE TITLE
Code compiles even with undefined references

### ISSUE THREAD

**Issue by tinloaf** [mattgodbolt/compiler-explorer#986]
While trying to construct a minimum non-working example of an "undefined reference" problem I'm seeing, I found that this code compiles in the compiler explorer (with any version of clang / gcc):

```c++
class Foo {
public:
    void method();
};

int main() {
    Foo f;
    f.method();
}
```

Obviously, `Foo::method()` in undefined, therefore this should trigger an undefined reference - error, right? It does if I compile this code locally with g++.

**Comment by partouf** [comment]
Undefined is triggered during linking only, one could have defined the implementation in another cpp file. If you click on binary mode (IOIOIO button) you’ll get this error.

But the default is to generate assembly without linking. Thus you don’t get the error.

**Comment by tinloaf** [comment]
What's the reason for not compiling to binary every time? Is it performance? Or the fact that the generated assembly looks nicer than the disassembled one? In the latter case, it might be useful to compile to binary even if the generated assembly is shown - to catch those undefined references and correctly show the error.

**Comment by mattgodbolt** [comment]
I very much prefer doing a partial compilation here, precisely as it means external symbols don't have to be provided, which is extremely handy when targeting a small area of code.

The reasons binary compilation isn't done is because the compilers themselves annotate their disassembly output, which is very useful both for compiler developers and people trying to understand their code. clang and icc both annotate branches with their predicted weights, for example. icc puts a lot of effort into annotations, e.g. https://godbolt.org/g/cdnxfp 

If we compile to binary we lose all that information, as well as open our users to all the useful (but confusing) dead code analysis.

In short, the code compiled is as if you had supplied `-c` to the compiler: proving your code links is not the main reason for Compiler Explorer, but looking at the generated code is.

>  In the latter case, it might be useful to compile to binary even if the generated assembly is shown - to catch those undefined references and correctly show the error.

As I say, I consider this an important feature, not a bug. I'm happy to discuss more, but I am closing this for now.
